ocaml/typing/typecore.ml

```diff
@@ -3733,13 +3757,12 @@ and type_approx_aux env sexp in_function ty_expected =
   match sexp.pexp_desc with
   | Pexp_let (_, _, e) -> type_approx_aux env e None ty_expected
   | Pexp_fun (l, _, p, e) ->
-      type_function_approx env sexp.pexp_loc l (Some p) (Psimple_rhs e)
-        in_function ty_expected
+      type_function_approx env sexp.pexp_loc l (Some p) e in_function
+        ty_expected
   | Pexp_function ({pc_rhs}::_) ->
       type_function_approx env sexp.pexp_loc Nolabel None pc_rhs
         in_function ty_expected
```

Comment:
I think this could be a bug: we're extracting out `pc_rhs` without checking whether it's Jane Syntax. `type_approx_*` is used for "approximately" type-checking all the bindings in a recursive knot before type-checking them for real &mdash; it's probably worth adding a test that involves `let rec f = function ...` where the function has a when-guard to see if this is currently a bug.

Reply:
Added such a test, which passes on this version of the code (although because I don't know what the result of `type_approx_*` is used for, it could be silently producing the wrong result).

I still suspect that the code is wrong though, so I fixed the bug (I believe) by instead getting out a parsed rhs and passing it around.